examples/cloudml-collaborative-filtering/trainer/utils.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2019 Google Inc. All Rights Reserved.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility functions for model training."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import random
+import tensorflow_transform as tft
+from google.cloud import storage
+
+from constants import constants  # pylint: disable=g-bad-import-order
+
+
+def _sample_vocab(tft_output, vocab_name, label, k):
+  """Samples the given vocab and returns the indices and samples.
+
+  Args:
+    tft_output: a TFTransformOutput object.
+    vocab_name: the name of the embedding vocabulary made with tft.
+    label: a label to assign each sample of the vocab.
+    k: the number of samples to take.
+
+  Returns:
+    A tuple of (indices, metadata):
+      indices: a list of indices for the vocab sample.
+      metadata: a list of lists of data corresponding to the indices.
+
+  Raises:
+    RuntimeError: k is larger than the vocab size.
+  """
+  vocab = tft_output.vocabulary_by_name(vocab_name)
+  if k > len(vocab):
+    raise RuntimeError("{0} num samples too high, must be at most {1}"
+                       .format(label, len(vocab)))
+
+  indices = random.sample(xrange(len(vocab)), k)
+  return indices, [[label, vocab[i]] for i in indices]
+
+
+def _write_projector_metadata_local(metadata_dir, metadata):
+  """Write local metadata file to use in tensorboard to visualize embeddings."""
+  metadata_path = os.path.join(metadata_dir, constants.PROJECTOR_PATH)
+  if not os.path.exists(metadata_dir):
+    os.makedirs(metadata_dir)
+  with open(metadata_path, "w+") as f:
+    f.write("label\tname\n")
+    f.write("\n".join(["\t".join(sample) for sample in metadata]))
+
+
+def _write_projector_metadata_gcs(metadata_dir, metadata):
+  """Write GCS metadata file to use in tensorboard to visualize embeddings."""
+  metadata_path = os.path.join(metadata_dir, constants.PROJECTOR_PATH)
+  split_path = metadata_path.split("/")
+  bucket_name = split_path[2]
```

Comment:
Lol I've run into this annoying gcs uri -> bucket / object things problem a bunch of times.
I think the cleanest way to do this is:
```python
bucket, obj_name = re.findall('gs:\/\/([^/]+)\/(.+)', gcs_uri)[0]
```

Or the ["official" way to do it from google cloud storage client library](https://github.com/googleapis/google-cloud-python/blob/master/storage/google/cloud/storage/client.py#L389).
```
from six.moves.urllib.parse import urlsplit
scheme, netloc, path, query, frag = urlsplit(blob_or_uri)
if scheme != "gs":
  raise ValueError("URI scheme must be gs")
```

Reply:
`netloc` will be your bucket name,  `path` will be your path.

Reply:
thanks for pointing this out!